hardware/Tesla.cpp

```diff
@@ -0,0 +1,502 @@
+#include "../main/localtime_r.h"
+#include "../main/json_helper.h"
+#include "../main/RFXtrx.h"
+#include "../main/SQLHelper.h"
+#include <sstream>
+#include <iomanip>
+#include <cmath>
+
+#define TESLA_SWITCH_CHARGE 1
+#define TESLA_SWITCH_CLIMATE 2
+#define TESLA_SWITCH_DEFROST 3
+
+#define TESLA_TEMP_CLIMATE 1
+#define TESLA_ALERT_STATUS 1
+#define TESLA_LEVEL_BATTERY 1
+
+#define TESLA_MAXTRIES 5
+#define TESLA_MAXIDLE 30
+
+CTesla::CTesla(const int ID, const std::string& username, const std::string& password, const std::string& VIN) :
+	m_username(username),
+	m_password(password)
+{
+	m_loggedin = false;
+	m_HwdID = ID;
+	m_car.VIN = VIN;
+	Init();
+	m_commands.clear();
+}
+
+CTesla::~CTesla(void)
+{
+}
+
+void CTesla::Init()
+{
+	m_car.charging = false;
+	m_car.connected = false;
+	m_car.is_home = false;
+	m_car.climate_on = false;
+	m_car.defrost = false;
+	m_car.awake = false;
+	m_trycounter = 0;
+}
+
+bool CTesla::Reset(std::string errortext)
+{
+	if(m_trycounter > TESLA_MAXTRIES)
+	{
+		Init();
+		SendSwitch(TESLA_SWITCH_CHARGE, 1, 255, m_car.charging, 0, m_Name + " Charge switch");
+		SendSwitch(TESLA_SWITCH_CLIMATE, 1, 255, m_car.climate_on, 0, m_Name + " Climate switch");
+		SendSwitch(TESLA_SWITCH_DEFROST, 1, 255, m_car.defrost, 0, m_Name + " Defrost switch");
+		m_commands.clear();
+		Log(LOG_ERROR, "Multiple tries requesting %s. Assuming car offline.", errortext.c_str());
+		SendAlertSensor(TESLA_ALERT_STATUS, 255, Offline, "Offline", m_Name + " State");
+		return(false);
+	}
+	else
+	{
+		Log(LOG_ERROR, "Timeout requesting %s. Ignoring.", errortext.c_str());
+		m_trycounter++;
+	}
+
+	return(true);
+}
+
+bool CTesla::StartHardware()
+{
+	RequestStart();
+
+	Init();
+	//Start worker thread
+	m_thread = std::make_shared<std::thread>(&CTesla::Do_Work, this);
+	SetThreadNameInt(m_thread->native_handle());
+	if (!m_thread)
+		return false;
+	m_bIsStarted = true;
+	sOnConnected(this);
+	return true;
+}
+
+bool CTesla::StopHardware()
+{
+	if (m_thread)
+	{
+		RequestStop();
+		m_thread->join();
+		m_thread.reset();
+	}
+	m_bIsStarted = false;
+	return true;
+}
+
+std::string CTesla::GetCommandString(const eCommandType command)
+{
+	switch (command)
+	{
+	case Send_Climate_Off:
+		return("Switch Climate off");
+	case Send_Climate_Comfort:
+		return("Switch Climate on");
+	case Send_Climate_Defrost:
+		return("Switch Defrost mode on");
+	case Send_Climate_Defrost_Off:
+		return("Switch Defrost mode off");
+	case Send_Charge_Start:
+		return("Start charging");
+	case Send_Charge_Stop:
+		return("Stop charging");
+	case Get_Climate:
+		return("Get Climate status");
+	case Get_Charge:
+		return("Get Charge status");
+	case Get_Drive:
+		return("Get Drive status");
+	default:
+		return "";
+	}
+}
+
+void CTesla::Do_Work()
+{
+	int sec_counter = 593;
+	int sec_counter_after_active = 0;
+	bool waking_up = false;
+	int interval = 1000;
+	Log(LOG_STATUS, "Worker started...");
+
+	while (!IsStopRequested(interval))
+	{
+		if (IsStopRequested(0))
```

Comment:
The second IsStopRequested is not needed.
When it's time to quit the worker, it stops directly